packages-nextgen/kestrel_core/src/kestrel/cache/sqlite.py

```
@@ -0,0 +1,151 @@
+        self.connection.execute(text(f'DROP TABLE "{instruction_id.hex}"'))
+
+    def _to_sql(self, df: DataFrame, name: str):
+        df.to_sql(name, con=self.connection, if_exists="replace", index=False)
+
+    def store(
+        self,
+        instruction_id: UUID,
+        data: DataFrame,
+    ):
+        self._to_sql(data, instruction_id.hex)
+
+    def evaluate_graph(
+        self,
+        graph: IRGraphSoleInterface,
+        instructions_to_evaluate: Optional[Iterable[Instruction]] = None,
+    ) -> Mapping[UUID, DataFrame]:
+        if not instructions_to_evaluate:
+            instructions_to_evaluate = graph.get_sink_nodes()
+        stacks = []
+        for inst in instructions_to_evaluate:
+            stacks.append(self._visit(graph, inst))
+        mapping = {}
+        for stack in stacks:
+            tail = stack[-1]
+            mapping[tail.id] = self._evaluate_instruction_stack(stack)
+        _logger.debug("mapping: %s", mapping)
+        return mapping
+
+    def _visit(
+        self,
+        graph: IRGraphSoleInterface,
+        instruction: Instruction,
+        stack: Optional[List[Instruction]] = None,
+    ) -> List[Instruction]:
+        """
+        Depth-first visit of nodes in graph from variable to variable or variable to source
+
+        Returns:
+            stack of Instructions visited
+        """
+        if not stack:
+            stack = []
+        # TODO: handle multiple predecessors of a node
+        _logger.debug("_visit: %s", instruction)
+        if isinstance(instruction, SourceInstruction) or (
+            isinstance(instruction, Variable) and len(stack) > 0
+        ):
+            # Need to first check if stack is empty since we start with var
+            # Since it's non-empty, this var should terminate the chain
+            stack.append(instruction)
+            return stack
+
+        stack = self._visit(graph, next(graph.predecessors(instruction)), stack)
```

Should we raise an exception for invalid graph given to evaluate such as non-SourceInstruction as 0-indegree node?

```
Filter -> ProjAttr
```

`graph.predecessors(instruction)` will return empty and `next()` will hit a `StopIteration` in the next call.